crates/apollo-parser/src/parser/mod.rs

```diff
@@ -210,9 +197,20 @@ impl Parser {
 
     /// Consume a token from the lexer.
     pub(crate) fn pop(&mut self) -> Token {
-        self.tokens
-            .pop()
-            .expect("Could not pop a token from the AST")
+        if let Some(token) = self.current_token.take() {
+            return token;
+        }
+
+        loop {
+            match self
+                .lexer
+                .next()
+                .expect("Could not pop a token from the AST")
+            {
+                LexerResult::Error(e) => self.errors.push(e),
+                LexerResult::Token(token) => return token,
+            }
+        }
```

Comment:
We should try to iterate outside of all the peek/pop functions. It might make it a tad difficult to debug later on when it's done in several places.